### DISCUSSION TITLE
Would you like changes to enable AIX shared libraries to use simpler packaging and support shlib_variant?

### DISCUSSION THREAD

**Discussion by johnkohl-hcl** [openssl/openssl#21504] [General Discussion]
Reading the discussion guidelines, I'm not sure this fits primarly as a discussion or an issue.  I'd welcome feedback.  If it's favorable, I can file an issue and a proposed fix.

AIX shared libraries can be packaged two ways:
* a shared object file inside an archive file, often described as `libxxx.a(foo.o)`: where `foo.o` is a shared object (this was the original support and the only option for some number of AIX releases)
* a bare shared object file, typically named `libxxx.so` (supported for several years now on supported AIX releases)

Shared libraries that depend on other libraries get their embedded dependency recorded based on how the linker found the library.  Example: if `foo.o` depends on `libY.a(bar.o)`, and is created with a link step that uses `-lY`, then the runtime dependency inside `foo.o` is to `libY.a(bar.o)`.  When `foo.o` is packed into the archive, then `libX.a(foo.o)` has a dependency on `libY.a(bar.o)`.  If `libX.so` depends on `libY.so`, and is created with a link step that uses `-lY`, then the runtime dependency inside `libX.so` is to `libY.so`.

OpenSSL builds AIX libraries as:
* `libssl.a(libssl.so)`: shared library in an archive
* `libcrypto.a(libcrypto.so)`: shared library in an archive
* `libssl_a.a`: static library
* `libcrypto_a.a`: static library

For OpenSSL, libssl depends on libcrypto, and is created by a linker command using `-lcrypto`.  The shared object for libssl will have an embedded reference to `libcrypto.a(libcrypto.so)`.

This all works OK with the OpenSSL builds until you attempt to configure a shlib_variant.

The OpenSSL model as I understand it is that for shlib_variant builds, the applications can still use the same include file names and link with `-lssl -lcrypto`, but the embedded references will be to the shared objects' variant names, such as `libssl-foo.so` and `libcrypto-foo.so` on other platforms.  You only need to install the variant-named libraries to get a working application.

On AIX, shlib_variant OpenSSL builds produce shared libraries like this (with variant "-foo"):
```
libcrypto.a(libcrypto-foo.so)
libssl.a(libssl-foo.so)
```

The application can link `-lssl -lcrypto`, but the runtime dependencies are a mixture of variant and non-variant names.  The containing archive name (`libcrypto.a`) is used for the reference, and thus you have to install the archive library with its original name (`libcrypto.a`). [Or, not supported by the build system, an administrator could modify an existing (system) library archive of the same name to add the variant-named shared object as an archive member.  I wouldn't recommend this, since it modifies some other package's files!]

The application depends on `libssl.a(libssl-foo.so)` and `libcrypto.a(libcrypto-foo.so)`.

If you try to cure the need for non-variant named archive files after an OpenSSL build, by renaming the shared library archives, to get these files/contents:
```
libssl-foo.a(libssl-foo.so)
libcrypto-foo.a(libcrypto-foo.so)
```
it only half-works when linking an application.  The application can be linked `-lssl-foo -lcrypto-foo`, and has embedded references to `libssl-foo.a(libssl-foo.so)` and `libcrypto-foo.a(libcrypto-foo.so)`.

But `libssl-foo.a(libssl-foo.so)` has a dependency on the shared library in the *original* archive name seen when `libssl-foo.so` was linked, namely `libcrypto.a(libcrypto-foo.so)`.  That means you still have to install `libcrypto.a` onto the target.  That breaks (for us) the model of wanting to install libraries only with file system names like `libcrypto-foo.a`.

Rather than deal with all the hassle of a shared object inside an archive, we can instead change the way OpenSSL packages libraries on AIX to use a shared object.

There's a way to create files such that `libssl.so.3` has a runtime embedded dependency on `libcrypto.so.3`:
* `libssl.so` : a link-import library that refers to libssl.so.3 (some simple scripting can create this during a build)
* `libcrypto.so` : a link-import library that refers to libcrypto.so.3
* `libssl.so.3`: the shared library
* `libcrypto.so.3`: the shared library

To link applications, you need both the `.so` and the `.so.3` for each library.  For example, when linking with `-lcrypto`, the linker processes `libcrypto.so`, sees its reference to `libcrypto.so.3`, and creates the application with a runtime library dependency on `libcrypto.so.3`.

At run time, the application directly references `libssl.so.3` and `libcrypto.so.3`.  `libssl.so.3` itself references `libcrypto.so.3`.

With this build style, shlib_variant now works OK.  You get:
* `libssl.so` : link-import library that refers to `libssl-foo.so.3`
* `libcrypto.so` : link-import library that refers to `libcrypto-foo.so.3`
* `libssl-foo.so.3`: the shared library
* `libcrypto-foo.so.3`: the shared library
* `libssl.a`: static library
* `libcrypto.a`: static library

At link time you still specify `-lssl -lcrypto`; at run time you only install `libssl-foo.so.3` and `libcrypto-foo.so.3`.

So, I can contribute changes to enable this style of shared library support on AIX.

1. Would you want such changes?
1. Would you want them to replace the existing AIX packaging (shlib-in-archive), or as a non-default alternative shared_target value (such as "aix-solib")?

**Comment by paulidale** [comment]
I think they'd be worthwhile.  I'd probably go for the non-default alternative rather than replacing the existing packaging (it's more compatible).

**Comment by johnkohl-hcl** [comment]
My company is using OpenSSL 3.0 series (long-term support branch) and eventually we'll want it there.  But to start out, which branch should I target for a PR?

**Reply by t8m** [reply]
This would be a new feature so it would not be eligible for 3.0 branch. Please target the master branch.

**Reply by johnkohl-hcl** [reply]
Thanks.  I filed issue #21518 with the feature request and will work up a pull request for it.

**Comment by makr** [comment]
Hm... I do the variant stuff on AIX, it creates the `lib*-variant.so.3` files. Later I do the linking in my application with `-bnoipath /path/to/openssl-libraries/libssl-variant.so.3 /path/to/openssl-libraries/libcrypto-variant.so.3`. From the top my head this is all what is necessary, but I would need to verify this on the actual build machine.

**Reply by johnkohl-hcl** [reply]
`-bnoipath` affects all libraries in the link, not just the particular libraries, so that seems undesirable as a workaround.  We're looking for the ability to link "normally" with `-lsss -lcrypto` and get run-time references to the variant names, like we get on other UNIXy platforms.

But also, look at the dependencies embedded inside `libssl-variant.so.3` with `dump -H` :
```
% dump -H -X64 libssl-variant.so.3 

libssl-variant.so.3:

			***Loader Section***
                      Loader Header Information
VERSION#         #SYMtableENT     #RELOCent        LENidSTR
0x00000001       0x00000403       0x00001692       0x00000060       

#IMPfilID        OFFidSTR         LENstrTBL        OFFstrTBL
0x00000003       0x0001c9a0       0x00005b7c       0x0001ca00       


			***Import File Strings***
INDEX  PATH                          BASE                MEMBER              
0      /opt/rational/common/shlib:/usr/lib:/lib                                         
1                                    libcrypto.a         libcrypto-variant.so.3
2                                    libc.a              shr_64.o            

```
at runtime, `libssl-variant.so.3` is going to try and load `libcrypto.a` from the file system in its attempt to find `libcrypto-variant.so.3` inside that file.  With variant naming, I don't want to be installing `libcrypto.a` for runtime use, it can conflict with some other `libcrypto.a`.  I want to be building such that only files named xxx-variant-yyy need to be installed.
